### ISSUE TITLE
fun with least-positive-single-float

### ISSUE THREAD

**Issue by kpoeck** [clasp-developers/clasp#1188]
Mentionned by scymtym for sbcl.

least-positive-single-float does not seem to be the least-positive-single-float

`(random least-positive-single-float)` should return the single-float 0.0 but doesn't
The result of `(random least-positive-single-float)` is still a single-float but smaller than `least-positive-single-float`

`least-positive-single-float` is defined as 
`cl::_sym_leastPositiveSingleFloat->defconstant(clasp_make_single_float(FLT_MIN));``

`FLT_MIN` is `1.17549e-38` on my machine as is `least-positive-single-float`

experimentally the real `least-positive-single-float` seems to be `1.4012985e-45` which the same value that ecl and sbal give on my machine.

All that newwest everything on MacOs BigSur

```lisp
COMMON-LISP-USER> least-positive-single-float
1.1754944e-38
COMMON-LISP-USER> (random least-positive-single-float)
6.968855e-39
COMMON-LISP-USER> (typep (random least-positive-single-float) 'single-float) 
T
COMMON-LISP-USER> (< (random least-positive-single-float) least-positive-single-float)
T
````

**Comment by Bike** [comment]
apparently `FLT_MIN` is the smallest positive _normalized_ float. The smallest positive float at all is the subnormal `FLT_TRUE_MIN`. Stupid but sounds easy to fix?

**Comment by kpoeck** [comment]
Trying your proposed solution now

**Comment by Bike** [comment]
Okay, so I checked the CLHS and I didn't actually know this, but CL makes the same distinction. We have `least-positive-single-float` and `least-positive-normalized-single-float`, etc. "normalized" is defined directly in reference to IEEE 754. It seems clear then that `least-positive-single-float` should be `FLT_TRUE_MIN` and `least-positive-normalized-single-float` should be `FLT_MIN`.

**Comment by kpoeck** [comment]
I believe in the definition of clasp drmeister simply mixed up the definitions of normalized and non normalized. I swiched that as you said in the above paragraph and this works better.

Still have the problem that (random x) should return a number lower than x, and (random least-positive-single-float9 returns least-positive-single-float.

to fix that I did - for the single-float case.
```c++
    float flimit = olimit.unsafe_single_float();
    if (flimit >  0.0f) {
      std::uniform_real_distribution<> range(0.0, flimit);
      float ran = range(random_state->_Producer._value);
      if (ran == flimit)
        return clasp_make_single_float(0.0f);
      else
        return clasp_make_single_float(ran);
`````
Seems to work, not sure if this is the best fix.

I tried to reconstruct that in a single c++ program, but that behaves even different.

Any advise?

**Comment by Bike** [comment]
This is apparently a common bug in C++ implementations; [here's LLVM's](https://bugs.llvm.org/show_bug.cgi?id=18767). It includes an example program using `nextafter` that shows basically the same problem, or I just wrote this quick one:

```c++
#include <limits>
#include <random>
#include <cstdio>

int main () {
  double f = std::numeric_limits<double>::denorm_min();
  printf("Denorm min: %f\n", f);

  std::uniform_real_distribution<double> dist(0.0, f);
  std::default_random_engine eng;

  for (int i = 0; i < 10; ++i) {
    double v = dist(eng);
    printf("Random: %f", v);
    if (f == v) printf(" (bug!)\n");
    else printf("\n");
  }
}
```

at least for the moment I'm not very worried about it. Returning zero probably screws up the distribution even more, so I don't know about doing that. The good solution would probably be to implement our own random floats, but I'm not sure I want to reinvent the wheel over this.

**Comment by kpoeck** [comment]
Well I thought 0 is better than than returning the argument passed to random

**Comment by kpoeck** [comment]
Just tried your example in xcode and got:
```c++
Denorm min: 0.000000
Random: 0.000000
Random: 0.000000 (bug!)
Random: 0.000000
Random: 0.000000
Random: 0.000000 (bug!)
Random: 0.000000 (bug!)
Random: 0.000000 (bug!)
Random: 0.000000 (bug!)
Random: 0.000000 (bug!)
Random: 0.000000 (bug!)
````

**Comment by kpoeck** [comment]
Another tought, if the argument to random is least-positive-single-float we could directly return 0.0. Same for the other float types-

Would that be better?

**Comment by Bike** [comment]
the C++ PRNG will also incorrectly return the high bound even for larger intervals like [0, 1), it's just less obvious since it doesn't happen as often.

**Comment by kpoeck** [comment]
So better just document the bug instead of imcomplete workarounds?

**Comment by Bike** [comment]
Yeah. How about we open up a separate bug for the problem with `random`, and in the meantime fix the float constants as you described

**Comment by kpoeck** [comment]
Will do
